src/analysis/normed_space/M_structure.lean

```diff
@@ -0,0 +1,380 @@
+  intro x,
+  refine le_antisymm _ _,
+  { calc ∥ x ∥ = ∥(P * Q) x + (x - (P * Q) x)∥ : by abel
+    ... ≤ ∥(P * Q) x∥ + ∥ x - (P * Q) x ∥ : by apply norm_add_le
+    ... = ∥(P * Q) x∥ + ∥(1 - P * Q) x∥ : rfl },
+  { calc ∥x∥ = ∥Q x∥ + ∥(1-Q) x∥ : by rw h₂.right x
+    ... = ∥P(Q x)∥ + ∥(1-P)(Q x)∥ + ∥(1-Q) x∥ : by rw h₁.right (Q x)
+    ... = ∥P(Q x)∥ + ∥Q x - P (Q x)∥ + ∥x - Q x∥ : rfl
+    ... = ∥P(Q x)∥ + (∥Q x - P (Q x)∥ + ∥x - Q x∥) : by rw add_assoc
+    ... ≥ ∥P(Q x)∥ + ∥(Q x - P (Q x)) + (x - Q x)∥ :
+      by apply (add_le_add_iff_left (∥P(Q x)∥)).mpr (norm_add_le (Q x - P (Q x)) (x - Q x))
+    ... = ∥P(Q x)∥ + ∥x - P (Q x)∥ : by rw sub_add_sub_cancel'
+    ... = ∥(P * Q) x∥ + ∥(1 - P * Q) x∥ : rfl }
+end
+
+lemma join {P Q: X →L[𝕜] X} (h₁ : is_Lprojection P) (h₂ : is_Lprojection Q) :
+  is_Lprojection (P + Q - P * Q) :=
+begin
+  have e1:  1 - (1 - P) * (1 - Q) = P + Q - P * Q :=
+  calc 1 - (1 - P) * (1 - Q) = 1 -(1 - Q - P * (1 - Q)) : by rw [sub_mul, one_mul]
+  ... = Q + P * (1 - Q) : by rw [sub_sub, sub_sub_self]
+  ... = P + Q - P * Q : by rw [mul_sub, mul_one, add_sub, add_comm],
+  rw [← e1, ← is_Lprojection.Lcomplement_iff],
+  exact is_Lprojection.mul (is_Lprojection.Lcomplement h₁) (is_Lprojection.Lcomplement h₂)
+end
+
+instance : has_compl { f : X →L[𝕜] X // is_Lprojection f } :=
+⟨λ P, ⟨1-P, P.prop.Lcomplement⟩⟩
+
+@[simp] lemma coe_compl (P : {P : X →L[𝕜] X // is_Lprojection P}) :
+  ↑(Pᶜ) = (1:X →L[𝕜] X) - ↑P := rfl
+
+instance : has_inf {P : X →L[𝕜] X // is_Lprojection P} :=
+⟨λ P Q, ⟨P * Q, P.prop.mul Q.prop⟩ ⟩
+
+@[simp] lemma coe_inf (P Q : {P : X →L[𝕜] X // is_Lprojection P}) :
+  ↑(P ⊓ Q) = ((↑P : (X →L[𝕜] X)) * ↑Q) := rfl
+
+instance : has_sup {P : X →L[𝕜] X // is_Lprojection P} :=
+⟨λ P Q, ⟨P + Q - P * Q, P.prop.join Q.prop⟩ ⟩
+
+@[simp] lemma coe_sup (P Q : {P : X →L[𝕜] X // is_Lprojection P}) :
+  ↑(P ⊔ Q) = ((↑P:X →L[𝕜] X) + ↑Q - ↑P * ↑Q) := rfl
+
+instance : has_sdiff {P : X →L[𝕜] X // is_Lprojection P} :=
+⟨λ P Q, ⟨P * (1-Q), by exact is_Lprojection.mul P.prop (is_Lprojection.Lcomplement Q.prop) ⟩⟩
+
+@[simp] lemma coe_sdiff (P Q : {P : X →L[𝕜] X // is_Lprojection P}) :
+  ↑(P \ Q) = (↑P:X →L[𝕜] X) * (1-↑Q) := rfl
+
+instance : partial_order {P : X →L[𝕜] X // is_Lprojection P} :=
+{ le := λ P Q, (↑P:X →L[𝕜] X) = ↑(P ⊓ Q),
+  le_refl := λ P,
+  begin
+    simp only [coe_inf],
+    rw [← sq, projection_def],
+    exact P.prop.left,
+  end,
+  le_trans := λ P Q R,
+  λ h₁ h₂, begin
+    simp only [coe_inf],
+    have e₁: ↑P = ↑P * ↑Q := h₁,
+    have e₂: ↑Q = ↑Q * ↑R := h₂,
+    exact
+    ( begin
+      nth_rewrite_rhs 0 e₁,
+      rw [mul_assoc, ← e₂, ← e₁],
+    end : (↑P:X →L[𝕜] X) = ↑P * ↑R)
+  end,
+  le_antisymm := λ P Q,
+  λ h₁ h₂, begin
+    have e₁: ↑P = ↑P * ↑Q := h₁,
+    have e₂: ↑Q = ↑Q * ↑P := h₂,
+    exact subtype.eq (by rw [e₁, commute.eq (Lproj_commute P.prop Q.prop), ← e₂] :
+      (↑P:X →L[𝕜] X) = ↑Q),
+  end, }
+
+instance : has_zero {P : X →L[𝕜] X // is_Lprojection P}  :=
```

Comment:
I think you want this lemma here. You can use it later instead of e₁ e₂ `rfl` statements.
```suggestion
lemma le_def (P Q : {P : X →L[𝕜] X // is_Lprojection P}) : P ≤ Q ↔ (P : X →L[𝕜] X) = ↑(P ⊓ Q) :=
iff.rfl

instance : has_zero {P : X →L[𝕜] X // is_Lprojection P}  :=
```
Alternatively, you may want the right-hand side to be `↑P * ↑Q`, I'm not sure.